### ISSUE TITLE
Support for additional properties

### ISSUE THREAD

**Issue by pschichtel** [Kotlin/kotlinx.serialization#1978]
It is sometimes the case that I want to parse an object with a few concrete properties and a bunch of additional properties.

OpenAPI 3's [Responses Object](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#responses-object) is an example for this. It has the fixed `default` default plus additional fields for various http statuses.

so this could look like:

```json
{
  "default": {...},
  "200": {...},
  "400": {...},
  "404": {...}
}
```

And it would be great to be able to parse this into something like:

```kotlin
@Serializable
data class Responses(val default: Response, statuses: Map<String, Response>)
```

Jackson has the `@JsonAnySetter` annotation which can be used to implement this.

OpenAPI/JSON Schema support specifying object schemas that have both `properties` and `additionalProperties`.

Currently I think the only solution would be a hand-written KSerializer for Responses.

**Comment by pschichtel** [comment]
for reference the KSerializer I wrote for my example:

```kotlin
object ResponsesSerializer : KSerializer<Responses> {
    private val mapSerializer = MapSerializer(String.serializer(), Response.serializer())
    private const val defaultFieldName = "default"

    override val descriptor: SerialDescriptor = object : SerialDescriptor by mapSerializer.descriptor {
        override val serialName: String = "Responses"
    }

    override fun serialize(encoder: Encoder, value: Responses) {
        val pairs = listOfNotNull(
            value.default?.let { Pair(defaultFieldName, it) },
        ) + value.statuses.map { (key, value) -> Pair(key.code.toString(), value) }

        mapSerializer.serialize(encoder, pairs.toMap())
    }

    override fun deserialize(decoder: Decoder): Responses {
        val map = mapSerializer.deserialize(decoder).toMutableMap()
        val default = map.remove(defaultFieldName)
        val statuses = map.mapKeys { (key, _) -> HttpStatusCode(key.toUShort()) }
        return Responses(default, statuses)
    }
}
```

**Comment by sandwwraith** [comment]
See https://github.com/Kotlin/kotlinx.serialization/issues/959#issuecomment-683881720

**Comment by pschichtel** [comment]
Interesting alternative to my serializer, but it still requires me to write a custom serializer doesn't it?

**Comment by sandwwraith** [comment]
Yes, it still requires one, but given that `JsonTransformingSerializer` is much easier to implement, I'm not sure that this feature needs special support

**Comment by pschichtel** [comment]
I guess "require" is a strong word here. Technically it is obviously not required since there is at least two alternative approaches.

However it would be great to have a dedicated feature for that, especially since this requires duplicating field names into the transformation which can easily go out of sync with the model.

An annotation `@AdditionalFields` or `@UnknownFields` on a `Map<*, *>` field that then automatically collects all fields that are not consumed by other fields would be what comes to my mind.

**Comment by efemoney** [comment]
> Yes, it still requires one, but given that `JsonTransformingSerializer` is much easier to implement, I'm not sure that this feature needs special support

The use cases for KxS are more than just JSON, would be great to have this possibly supported in custom formats

**Comment by svenjacobs** [comment]
This would still be very much appreciated :) 

I'm trying to deserialize the payload part of a [JWT](https://jwt.io). A JWT contains some standardized fields and a lot of possible [optional fields](https://www.iana.org/assignments/jwt/jwt.xhtml). I would like to put all unknown fields into a `Map<String, String>`. Something like

```kotlin
@Serializable
@JsonUnknownKeysInto
val unknown: Map<String, String>
```

**Comment by svenjacobs** [comment]
This would still be very much appreciated 😃 

I'm currently trying to deserialize the payload of a [JWT](https://jwt.io). JWTs include a set of standardized fields, but they can also contain many [optional fields](https://www.iana.org/assignments/jwt/jwt.xhtml). My goal is to collect all unknown fields into a `Map<String, String>`. Ideally, it would be great to have something like:

```kotlin
@Serializable
@JsonUnknownKeysInto
val unknown: Map<String, String>
```

to handle this automatically!

**Comment by inemtsev** [comment]
I’d love to see first-class support for this, especially for the JSON use case.

It feels like there’s a missing middle ground between:

- failing on unknown keys, and
- silently dropping them with ignoreUnknownKeys = true

What would be really useful is a way to capture unknown JSON object members into a dedicated property, for example:

```kotlin
@Serializable
data class Payload(
    val known: String,
    @JsonExtraProperties
    val extra: Map<String, JsonElement> = emptyMap()
)
```

So that:
- known keys decode normally
- unknown keys are collected into extra
- nested values are preserved as JsonElement

If this makes sense, I can try creating a draft for implementation. I am thinking something like this:

- if @JsonExtraProperties exists, unknown keys are captured there
- ignoreUnknownKeys becomes irrelevant for that class’s unknown keys, because they are no longer “unknown” in practice
- if no extras property exists, current behavior stays unchanged
- nested unknown values are stored as raw JsonElement
- duplicate JSON keys follow current parser behavior, whatever the parser already does

**Comment by pdvrieze** [comment]
@inemtsev XML has support for this (as XML Schema explicitly supports "extra" attributes/children), although it is more complicated (there are elements, body content and attributes).

**Comment by inemtsev** [comment]
@sandwwraith I think I might have an efficient way of doing this. But, it requires adding a feature to the compiler plugin first. Can you take a look if I make a PR?

**Comment by pdvrieze** [comment]
@inemtsev You may want to first look at doing it outside the compiler plugin. I'm not sure either where the efficiency is. The only bit is in the descriptor where you may want to record this special child index (the base class is in the library so you can do it there).

**Comment by inemtsev** [comment]
> [@inemtsev](https://github.com/inemtsev) You may want to first look at doing it outside the compiler plugin. I'm not sure either where the efficiency is. The only bit is in the descriptor where you may want to record this special child index (the base class is in the library so you can do it there).

I couldn't find an efficient way to do it outside the compiler plugin, since deserializer is essentially created at compile-time for this library, we need to define a **sink** to put unknown fields at compile time. Otherwise, I have to parse the whole tree before deciding to collect unknowns which would kill performance for that class. I'll add a bit more details after work.

**Comment by pdvrieze** [comment]
> I couldn't find an efficient way to do it outside the compiler plugin, since deserializer is essentially created at compile-time for this library, we need to define a **sink** to put unknown fields at compile time. Otherwise, I have to parse the whole tree before deciding to collect unknowns which would kill performance for that class. I'll add a bit more details after work.

What happens in deserialization for (most, especially textual) formats is that the serializer calls `decodeElementIndex` in the format's `CompositeDecoder`. This function maps the name of the attribute to the index of the corresponding element as described in the containing struct's `SerialDescriptor`. For generated serializers this inherits from [`PluginGeneratedSerialDescriptor`](https://github.com/Kotlin/kotlinx.serialization/blob/master/core/commonMain/src/kotlinx/serialization/internal/PluginGeneratedSerialDescriptor.kt) and this is implemented as a map lookup. If no element index is available for the name this returns the UNKNOWN_NAME contstant that triggers unknown child behaviour in the format.

The way to have a fallback is to have one child have an appropriate annotation (and type), and then store the element content as generic JsonElement into the storage for that child.

The biggest performance "issue" is to determine which child to store unknown children in. This can be done by iterating over all elements, but would be faster if pre-computed. To precompute it, you would need to first add an appropriate property to `PluginGeneratedSerialDescriptor` (and maybe `CachedNames`) and then in `buildIndices` add a little check for the annotation and set the property (hopefully also detecting multiple annotations).

If you want to make it really fancy you may have a mechanism to just add format specific additional information to a descriptor. We don't want to really have format specific elements in this base type, so some sort or generalisation would be good. But note that this issue would still be true if you did things in the plugin.

As a suggestion, perhaps it would be an option to have an "indexedAnnotations" map where serialInfo annotations with a specific annotation (e.g. `IndexedSerialInfo`) would be added to it on descriptor creation. This would make matters generalizable while also allowing for map based lookup of "special" children.

**Comment by sandwwraith** [comment]
Indeed, having plugin support makes sense only if we want this feature to be universal. For Json alone, it can be done in the runtime.
It was also recently done for Protobuf, you can look at how it is implemented there.
Protobuf unknown fields though bring us back to the question " if we want this feature to be universal"......

**Comment by inemtsev** [comment]
OK, thanks for the tips guys, I think I see a way to do it directly for JSON. Just a quick behavioral question. Let's say I mark a field called "extras" with my annotation @JsonExtraKeys.  But there is an actual, JSON key with the name "extras", what should the behavior here be?

**Comment by pdvrieze** [comment]
@inemtsev The "additional keys" field should effectively ignore the field name. The main thing is that such an extras field must be of a narrow set of types (most likely just `Map<String,JsonElement>`) and you then have to deal with the question what the type of "extras" should be? Should the content of extras be stored as "extra" key inside the map, or should it replace the entire map. The latter is surprising and against the whole purpose of the feature.

As to making the feature universal. It hits the problem of how you represent this. Formats (and their parsing) are driven by the serializers, so storing "decoding events" is not possible in a general case. Too much is format specific (in XML there are two separate annotations, one for tag content, and one for attributes - it makes sense for XML, but would not for JSON.).

**Comment by inemtsev** [comment]
Yes, I mostly agree. So, I am leaning towards "extras" key inside the map `Map<String, JsonElement>` in case of conflict. Hopefully, this won't complicate implementation much. I'll try to share a PR this week.
